src/reshard/app/src/service.rs

```diff
@@ -0,0 +1,233 @@
+	/// and to cross-reference previous ceremonies for the same key.
+	#[serde(with = "qos_hex::serde")]
+	pub quorum_public_key: Vec<u8>,
+
+	/// Raw AWS Nitro **attestation document** bytes (as produced by NSM).
+	///
+	/// Share holders verify this before posting shares. The document binds:
+	/// - the enclave's measured state (PCRs / EIF),
+	/// - the **ephemeral public key** created at boot (used below in `signature`),
+	/// - the manifest hash via `user_data`.
+	#[serde(with = "qos_hex::serde")]
+	pub attestation_doc: Vec<u8>,
+
+	/// Envelope that **encapsulates the manifest and its approvals**, including:
+	///
+	/// - `manifest`
+	/// - `manifest_approvals`
+	/// - `share_set_approvals`
+	///
+	pub manifest_envelope: ManifestEnvelope,
+
+	/// Per-new-member outputs of the resharding step (**reuses genesis format**).
+	///
+	/// Each entry contains:
+	/// - the **member’s public key** in the **new share-set**,
+	/// - the **encrypted quorum key share** for that member,
+	/// - a **share hash** used to validate correct decryption **offline**.
+	pub member_outputs: Vec<GenesisMemberOutput>,
+
+	/// Ephemeral-key signature binding outputs to this **attested run**.
+	///
+	/// The ephemeral public key is carried in `attestation_doc`. The signature
+	/// is computed over `sha512(borsh(member_outputs))`. Verifiers should:
+	/// 1) parse & verify the attestation (incl. ephemeral pubkey),
+	/// 2) recompute the digest from `member_outputs`,
+	/// 3) verify this signature with the ephemeral pubkey.
+	#[serde(with = "qos_hex::serde")]
+	pub signature: Vec<u8>,
+}
+
+pub struct ReshardProcessor {
+	cached_reshard_bundle: ReshardBundle,
+}
+
+impl ReshardProcessor {
+	pub fn new<N: qos_nsm::NsmProvider + ?Sized>(
+		handles: &handles::Handles,
+		new_share_set: &ShareSet,
+		nsm: &N,
+	) -> Result<Self, String> {
+		// load keys
+		let quorum_pair: P256Pair = handles
+			.get_quorum_key()
+			.map_err(|e| format!("unable to get quorum key: {e:?}"))?;
+		let eph_pair: P256Pair = handles
+			.get_ephemeral_key()
+			.map_err(|e| format!("unable to get ephemeral key: {e:?}"))?;
+
+		let quorum_pub = quorum_pair.public_key().to_bytes();
+		let master_seed = quorum_pair.to_master_seed();
+
+		// Get attestation doc, which ties the running of this specific instance with:
+		// 1. the creation of eph key
+		// 2. the manifest and approvals
+		let attestation_doc =
+			match nsm.nsm_process_request(NsmRequest::Attestation {
+				user_data: Some(
+					handles.get_manifest_envelope().qos_hash().to_vec(),
+				),
+				nonce: None,
+				public_key: Some(eph_pair.public_key().to_bytes()),
+			}) {
+				NsmResponse::Attestation { document } => document,
+				other => {
+					return Err(format!("unexpected NSM response: {other:?}"))
+				}
+			};
+
+		// Split the master seed
+		let n = new_share_set.members.len();
+		let k = new_share_set.threshold as usize;
+		// shares_generate -> Result<Vec<Vec<u8>>, _>;
+		let shares: Vec<Vec<u8>> =
```

Comment:
```suggestion
		let shares =
```
Should be implied by the function signature return type
I would also delete the doc line above